### ISSUE TITLE
Any way to get Weights of Inception V4 models

### ISSUE THREAD

**Issue by pGit1** [titu1994/Keras-Classification-Models#1]
Hi there. It appears that it would be awesome if we could some how the weights of the inception v4 models similar to the weights of the other models you have built. Is this possible?

Here is a path to some weights although it is not exhaustive (looks like only Inception_ResNet_V2 is on there:

https://research.googleblog.com/2016/08/improving-inception-and-image.html

Not sure if this weight path can help either and work with the models you have created (seems like a more complete list):

https://github.com/tensorflow/models/blob/master/slim/README.md

**Comment by kentsommer** [comment]
@pGit1 

I have an implementation of Inception V4 with ported weights from the TFSlim implementation here: https://github.com/kentsommer/keras-inceptionV4

**Comment by pGit1** [comment]
Thanks Kent you are awesome!

**Comment by titu1994** [comment]
@kentsommer Can I ask whether these weights are for Tensorflow backend with TF dim ordering? Or for Theano backend with TH dim ordering?

The reason I ask is that if this is the weights for the Tensorflow backend with TF dim ordering, the weights dont automatically translate for Theano backend with TH dim ordering.

**Comment by kentsommer** [comment]
@titu1994 

You are correct the weights I have at the moment are only TF dim ordering with Tensorflow backend. However, once I return from my holiday (1week from now) I will be working on transforming them for the Theano backend and TH dim ordering.

All that being said the conversion for the way Theano handles convolution is very straightforward.

**Comment by titu1994** [comment]
@kentsommer Thats fantastic. There is one problem which I must mention. 

According to fcholet, when transposing the weights of the theano weights to tensorflow (th dim ordering to tf dim ordering) or tensorflow weights to theano (tf dim ordering to th dim ordering), not only must you transpose the weights of the convolution layer, but also shuffle the rows of the 1st dense layer after the last convolution layer to match the dim shuffle of the weights we just did.

Otherwise, while the tensorflow weights with TF dim ordering get correct accuracy and performance, the theano backend weights with th dim ordering will get extremely poor performance / not be able to classify anything at all.  The discussion is on the google group for Keras : https://groups.google.com/forum/#!searchin/keras-users/convert$20weights$20tensorflow%7Csort:relevance/keras-users/E1W4HpuxxFw/B2DCDluTCwAJ

If you are able to correctly translate the weights, please do share the code to do so in a gist. I am quite interested in how it is done, since i have several weights of various models trained in theano, but I can't correctly port them to tensorflow cause of above mentioned problem (I can't seem to understand how to shuffle the rows of the 1st dense layer after the last convolution layer).

**Comment by kentsommer** [comment]
@titu1994 

The conversion for this ended up not requiring a row-shuffle of the fully connected layer. However, if you do find it is necessary for the models you are trying to convert, [this gist](https://gist.github.com/kentsommer/e872f65926f1a607b94c2b464a63d0d3) may be useful...

Do note that I am not positive that this is correct, however, it seems to be what @fchollet did for VGG16 and VGG19 

One last thing... I found that trying to do the kernel conversion and transpose in one go did not work. I needed to split it into two separate python scripts. See the "convert_kernels.py" and "convert_dims.py" [here](https://github.com/kentsommer/keras-inceptionV4/tree/master/convert_weights) to see how I ended up making the conversion.

Feel free to shoot me an email (or just comment on the gist) if you have any questions!

**Comment by titu1994** [comment]
You are a saviour @kentsommer! I will definitely test the scripts out on the models I have to see if they correctly port the weights. Thank you so very much!

Also, it would be good to add the Inception V4 model to the Keras masker branch as an application.

**Comment by kentsommer** [comment]
@titu1994 

>  it would be good to add the Inception V4 model to the Keras masker branch as an application

That is the plan currently. Just need to get the model in line with PEP8 standards first.

**Comment by ashavish** [comment]
@kentsommer  awesome to see Inception-v4 on Keras. How can i get the layer outputs with include_top=False like it was in Inception V3 Keras implementation ?

**Comment by titu1994** [comment]
You can download the weights, use HDFView to delete the last 3 2-3 layers (Dense, Flatten, Pooling if need be) and then save the weights. 

Then take either of our model building scripts and delete the last 3 lines, add your layers and create the model, freeze all inception weights and load it's weights by name.

**Comment by kentsommer** [comment]
@ashavish 

Follow what @titu1994 suggested, however, it should not be necessary to delete the layers from the weights files. Since the weights are loaded by name, the weights for the layers that are no longer in your model definition will simply not be used for anything. 

Also @ashavish if you open an issue on my model repository (I'm sure @titu1994 would do the same as well), once I have some free time I will add the include_top definition similar with the current Keras applications.

**Comment by ashavish** [comment]
Thanks @kentsommer , @titu1994 , already implemented it by using the getting intermediate layers method... base_model.get_layer('merge_25').output . Hopefully this should also yield the same results.
